### ISSUE TITLE
Error when running fine-tuning job with SageMaker

### ISSUE THREAD

**Issue by mmcclean-aws** [huggingface/optimum-neuron#41]
Am trying to use a custom cache repo with a SageMaker training job and getting an error message

I have set the following env vars including `HUGGING_FACE_HUB_TOKEN` and `CUSTOM_CACHE_REPO`

```
Traceback (most recent call last):
  File "run_glue.py", line 622, in <module>
main()
  File "run_glue.py", line 530, in main
train_result = trainer.train(resume_from_checkpoint=checkpoint)
  File "/usr/local/lib/python3.8/site-packages/transformers/trainer.py", line 1662, in train
return inner_training_loop(
  File "/usr/local/lib/python3.8/site-packages/transformers/trainer.py", line 1929, in _inner_training_loop
tr_loss_step = self.training_step(model, inputs)
  File "/usr/local/lib/python3.8/site-packages/transformers/trainer.py", line 2699, in training_step
loss = self.compute_loss(model, inputs)
  File "/usr/local/lib/python3.8/site-packages/optimum_neuron-0.1.1.dev0-py3.8.egg/optimum/neuron/trainers.py", line 180, in compute_loss
self.trigger_on_step_middle_for_neuron_cache_callback(model)
  File "/usr/local/lib/python3.8/site-packages/optimum_neuron-0.1.1.dev0-py3.8.egg/optimum/neuron/trainers.py", line 176, in trigger_on_step_middle_for_neuron_cache_callback
callback.on_step_middle(self.args, self.state, self.control, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/optimum_neuron-0.1.1.dev0-py3.8.egg/optimum/neuron/trainer_callback.py", line 260, in on_step_middle
    self.neuron_hash_for_model(args, model, state.last_inputs, try_to_fetch_cached_model=True)
  File "/usr/local/lib/python3.8/site-packages/optimum_neuron-0.1.1.dev0-py3.8.egg/optimum/neuron/trainer_callback.py", line 194, in neuron_hash_for_model
self.try_to_fetch_cached_model(neuron_hash)
  File "/usr/local/lib/python3.8/site-packages/optimum_neuron-0.1.1.dev0-py3.8.egg/optimum/neuron/trainer_callback.py", line 209, in try_to_fetch_cached_model
    found_in_cache = download_cached_model_from_hub(
  File "/usr/local/lib/python3.8/site-packages/optimum_neuron-0.1.1.dev0-py3.8.egg/optimum/neuron/utils/cache_utils.py", line 343, in download_cached_model_from_hub
huggingface_hub.snapshot_download(
  File "/usr/local/lib/python3.8/site-packages/huggingface_hub/utils/_validators.py", line 120, in _inner_fn
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/huggingface_hub/_snapshot_download.py", line 235, in snapshot_download
thread_map(
  File "/usr/local/lib/python3.8/site-packages/tqdm/contrib/concurrent.py", line 69, in thread_map
    return _executor_map(ThreadPoolExecutor, fn, *iterables, **tqdm_kwargs)
  File "/usr/local/lib/python3.8/site-packages/tqdm/contrib/concurrent.py", line 51, in _executor_map
    return list(tqdm_class(ex.map(fn, *iterables, chunksize=chunksize), **kwargs))
  File "/usr/local/lib/python3.8/site-packages/tqdm/std.py", line 1178, in __iter__
for obj in iterable:
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 619, in result_iterator
yield fs.pop().result()
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 444, in result
return self.__get_result()
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
raise self._exception
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.8/site-packages/huggingface_hub/_snapshot_download.py", line 212, in _inner_hf_hub_download
    return hf_hub_download(
  File "/usr/local/lib/python3.8/site-packages/huggingface_hub/utils/_validators.py", line 120, in _inner_fn
return fn(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/huggingface_hub/file_download.py", line 1331, in hf_hub_download
_chmod_and_replace(temp_file.name, local_dir_filepath)
  File "/usr/local/lib/python3.8/site-packages/huggingface_hub/file_download.py", line 1523, in _chmod_and_replace
os.replace(src, dst)
OSError
: [Errno 18] Invalid cross-device link: '/root/.cache/huggingface/hub/tmpr2ar15_7' -> '/tmp/tmpuux64xpc/neuron-compile-cache/2.5.0.28+1be23f232/bert/062bef7185056b3ee894962abf656a6148b5898c28bc621100bba24ff2fd09284dbe361636bd5a7a8fcf200a1a15b5ac40798566dc283128b7d94d6a08598eeb/2dab9e373e1e65b6af81162e55c980514f719456fcb0c5f32b1d5d25d408f6856ba2acfc802b233521ae12127255ce13c1032ba295504279ac70fb87b2d93af2/MODULE_11427876144236435996/MODULE_6_SyncTensorsGraph.5983_11427876144236435996_algo-1-dd310922-102-5f9f28b6005cf/99f9bebb-2a08-4dc9-a9e1-c00816029eb3/MODULE_6_SyncTensorsGraph.5983_11427876144236435996_algo-1-dd310922-102-5f9f28b6005cf.neff'
```
Could it be because of the command `os.replace` as the source and destination are on different file systems ?

**Comment by philschmid** [comment]
I researched, and the error might come from `/root` not being on the same partition as `/tmp`. 

Can you try providing `HUGGINGFACE_HUB_CACHE=/tmp/hub`  env var when launching the training to see if it works?

**Comment by mmcclean-aws** [comment]
I believe this PR fixes the issue https://github.com/huggingface/huggingface_hub/pull/1439

**Comment by philschmid** [comment]
Yes it is already merged https://github.com/huggingface/optimum-neuron/pull/46 and released in `0.0.3`
